Deal with uninitialized variables in MOSART_heat_mod.F90

liho745 opened the pull request:
The use of uninitialized variables (i.e., `Qsur` and `Qsub`) in the computation of the change of energy 
due to heat exchange with the environment and advection is fixed. Currently, it is assumed that those
changes in energy are zero.

Fixes #5332
[BFB]

ambrad reviewed (commented):
This resolves #5332. I don't have any knowledge of the meaning of the model changes, but my understanding is that these changes do not affect E3SM productions configurations. That is, this is a BFB PR from the perspective of E3SM, right?

liho745 commented:
> This resolves #5332. I don't have any knowledge of the meaning of the model changes, but my understanding is that these changes do not affect E3SM productions configurations. That is, this is a BFB PR from the perspective of E3SM, right?

I'd say yes based on my understanding, but I have not done any BFB test.

ambrad reviewed (approved):
I'll approve based on resolution of the issue. I expect this PR to be BFB. @bishtgautam, I suggest you run e3sm_developer on, e.g., Chrysalis to be sure that it is indeed BFB before merging it to next.

bishtgautam reviewed (changes requested):
Minor code clean up changes requested.

bishtgautam commented:
@ambrad Typically, the developer is responsible for running the regression test suite and ensuring the test suite's results are as expected.

liho745 commented:
> @ambrad Typically, the developer is responsible for running the regression test suite and ensuring the test suite's results are as expected.

I have addressed the comments from @bishtgautam, and performed the E3SM land developer's tests on Compy -- all passed with BFB.

bishtgautam commented:
@liho745 I have edited the first comment of this PR, which will be copied/pasted as the merge commit message. For the record, below is the original first comment

> In MOSART_heat_mod.F90, Qsur and Qsub were not initialized before using them in Line 91. Lines 91 and 92 calculate deltaM_t and deltaH_t that are actually not used in the later calculations, i.e., since here a simplified approach is used to calculate the stream temperature instead of the energy balance approach. I commented off Qsur and Qsub and made another couple of minor changes. A global test run was performed with the IELM compset and r05_r05 grids, and heatflag was set to true. The simulated stream temperature seems reasonable in terms of magnitude and spatial patterns.

bishtgautam commented:
@liho745 How about also adding a test for this stealth feature?

liho745 commented:
> @liho745 I have edited the first comment of this PR, which will be copied/pasted as the merge commit message. For the record, below is the original first comment
> 
> > In MOSART_heat_mod.F90, Qsur and Qsub were not initialized before using them in Line 91. Lines 91 and 92 calculate deltaM_t and deltaH_t that are actually not used in the later calculations, i.e., since here a simplified approach is used to calculate the stream temperature instead of the energy balance approach. I commented off Qsur and Qsub and made another couple of minor changes. A global test run was performed with the IELM compset and r05_r05 grids, and heatflag was set to true. The simulated stream temperature seems reasonable in terms of magnitude and spatial patterns.

Sounds good.

liho745 commented:
> @liho745 How about also adding a test for this stealth feature?

I don't quite follow. Can you elaborate?

bishtgautam commented:
Add a new test (say `ERS.<RES>.<COMPSET>.mosart-heat`) in the `e3sm_mosart_developer` test suite in [tests.py#L13](https://github.com/E3SM-Project/E3SM/blob/master/cime_config/tests.py#L13) and add a corresponding new `heat` directory in [`components/mosart/cime_config/testdefs/testmods_dirs/mosart`](https://github.com/E3SM-Project/E3SM/tree/master/components/mosart/cime_config/testdefs/testmods_dirs/mosart). Does that make sense?

hydrotian commented:
@liho745 I wrote a [document](http://simhydro.com/notebook//e3sm/programming/2021/07/29/E3SM-user-defined-test.htm) about adding a new ELM test last year. Hope the information is still up to date: l

liho745 commented:
> 

Many thanks! I have added a new test for MOSART-heat, and passed the baseline tests BFB.

bishtgautam commented on cime_config/tests.py:
```diff
@@ -13,13 +13,21 @@
     "e3sm_mosart_developer" : {
         "share" : True,
         "time"  : "0:45:00",
+        "inherit" : ("e3sm_mosart_heat"),
```
@liho745 Thanks for adding the test. Instead of creating a new test suite (i.e. `e3sm_mosart_heat`), please directly add the new test in the `e3sm_mosart_developer` test suite at https://github.com/E3SM-Project/E3SM/pull/5343/files#diff-e2fbf220fd98012c3bdf61d7f341a1815f43823af2ffd3bf926d5aafc4363d8bR20. Does it make sense?

bishtgautam replied:
I will have to revisit the approach taken for the sediment test when I re-merge that PR into `next`. To avoid creating many new test suites, directly add `ERS.r05_r05.RMOSGPCC.mosart-heat` in the `e3sm_mosart_developer` test suite.

bishtgautam replied:
>I quote from him "mosart-sediment cannot run with other test cases with the same build".

Okay, I now recall this issue with mosart sediment. Does your mosart heat test need different build flags (not namelist flags) than those tests in the e3sm_mosart_developer test suite?

liho745 replied:
Yes, I have to turn on heatflag via user_nl_mosart.

bishtgautam replied:
That is a namelist flag and not a build flag (e.g. `-DUSE_PETSC_LIB`), which is used when compiling various `*.F90`. Thus, you should be able to directly add the new test in the e3sm_mosart_developer test suite.

liho745 replied:
> That is a namelist flag and not a build flag (e.g. `-DUSE_PETSC_LIB`), which is used when compiling various `*.F90`. Thus, you should be able to directly add the new test in the e3sm_mosart_developer test suite.

I tried what you said, and the test results are below, i.e., the baseline test failed. 
====================
  ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat (Overall: DIFF) details:
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat CREATE_NEWCASE
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat XML
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat SETUP
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat SHAREDLIB_BUILD time=277
    FAIL ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat NLCOMP
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat MODEL_BUILD time=212
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat SUBMIT
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat RUN time=51
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat COMPARE_base_rest
    FAIL ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat BASELINE 21ffb4d: ERROR BFAIL baseline directory '/compyfs/liho623/e3sm_scratch/e3sm_baseline/21ffb4d/ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat' does not exist
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat MEMLEAK
    PASS ERS.r05_r05.RMOSGPCC.compy_intel.mosart-heat SHORT_TERM_ARCHIVER

bishtgautam replied:
The above failure is okay because the baseline for this new test doesn't exist.

liho745 replied:
Sounds good. Added the mosart-heat test directly into the e3sm_mosart_developer suite.
